sql/core/src/main/scala/org/apache/spark/sql/execution/aggregate/HashAggregateExec.scala

```diff
@@ -73,7 +73,9 @@ case class HashAggregateExec(
   // This is for testing. We force TungstenAggregationIterator to fall back to the unsafe row hash
   // map and/or the sort-based aggregation once it has processed a given number of input rows.
   private val testFallbackStartsAt: Option[(Int, Int)] = {
-    sqlContext.getConf("spark.sql.TungstenAggregate.testFallbackStartsAt", null) match {
+    Option(sqlContext).map { sc =>
```

Comment:
This is a hidden bug. `SubqueryExpression` will be sent to the executor side and build `Projection`, and be put in `EquivalentExpressions`, which needs to call `canonicalized`.

This means, Spark may serialize and send `HashAggregateExec` to the executor side, where `sqlContext` should be null.

It's hidden for a long time because `ScalarSubquery` didn't implement `canonicalized`, so the bug is not triggered. However, it also means `semanticHash` is wrong.

I think it only affects common subquery elimination, and shouldn't be a serious bug.

Reply:
In the long term, I think we should only send an "expression evaluator" to the executor side. The semantic check should only be done in the driver side.

Reply:
Sorry, for the late comment @cloud-fan. I think I've run into this issue before:
https://github.com/apache/spark/pull/28885/files#diff-9b62cef6bfdeb6c802bb120c7a724a974d5067a69585285bebb64c48603f8d6fR105-R108. The point is that there might be other nodes where canonicalization on executor side can cause issues. `SortExec.enableRadixSort` is the other one I found.

Reply:
`InSubqueryExec` already implements `canonicalized` before this PR, so we need to fix these bugs anyway.

I have an idea to fix this problem in all physical plans:
1. remove `SparkPlan.sqlContext`, so that we can catch all the callers of it
2. add `@transient final val session = SparkSession.getActiveSession.orNull`, to replace the previous `sqlContext`
3. override `conf` in `SparkPlan`: `if (session != null) session.sessionState.conf else SQLConf.get`

@peter-toth what do you think? AFAIK the only reason to access `SparkPlan.sqlContext` at executor side is to read a conf, and we can do that with `SQLConf.get` at executor side.

Reply:
Sounds good to me.

Reply:
I'm quite busy this week and may not have time to implement this idea recently. @peter-toth feel free to pick up this idea and open a PR if you have time, or I'll do it next or next next week. Thanks in advance!

Reply:
Ok, thanks. I will try to open a PR this week.

Reply:
Opened https://github.com/apache/spark/pull/32947.